src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Unix.cs

```diff
@@ -137,14 +138,22 @@ internal SocketError ReplaceHandle()
 
             // Then replace the handle with a new one
             SafeSocketHandle oldHandle = _handle;
-            SocketError errorCode = SocketPal.CreateSocket(_addressFamily, _socketType, _protocolType, out _handle);
+            SocketError errorCode = SocketPal.CreateSocket(_addressFamily, _socketType, _protocolType, out SafeSocketHandle newHandle);
+            Volatile.Write(ref _handle, newHandle);
             oldHandle.TransferTrackedState(_handle);
             oldHandle.Dispose();
+
             if (errorCode != SocketError.Success)
             {
                 return errorCode;
             }
 
+            if (Volatile.Read(ref _disposed) != 0)
+            {
+                _handle.Dispose();
+                ThrowIfDisposed();
```

Comment:
Nit: we know we are disposed at this point, so this seems more reasonable:
```suggestion
                throw new ObjectDisposedException(GetType().FullName);
```